src/requirements/graph/processor/add-arbitrary-opt-in-choices.ts

```diff
@@ -0,0 +1,47 @@
+import GraphProcessor from './interface';
+import RequirementFulfillmentGraph from '..';
+import { CourseWithUniqueId } from '../types';
+
+export type AddArbitraryOptInChoicesParameters<Requirement extends string> = {
+  /**
+   * The mapping from course's unique ID to arbitrary opt in choices.
+   */
+  readonly userChoiceOnRequirementOverrides: {
+    readonly [uniqueId: string]: {
+      readonly arbitraryOptIn: readonly Requirement[];
+    };
+  };
+};
+
+/**
+ * Add the user's arbitrary opt in choices to the graph.
+ */
+export default class AddArbitraryOptInChoices<
+  Requirement extends string,
+  Course extends CourseWithUniqueId
+> implements GraphProcessor<Requirement, Course> {
+  private arbitraryOptInChoicesParameters: AddArbitraryOptInChoicesParameters<Requirement>;
+
+  constructor(arbitraryOptInChoicesParameters: AddArbitraryOptInChoicesParameters<Requirement>) {
+    this.arbitraryOptInChoicesParameters = arbitraryOptInChoicesParameters;
+  }
+
+  public process(graph: RequirementFulfillmentGraph<Requirement, Course>) {
```

Comment:
Is it possible to indicate at the type level whether each operation mutates the graph? I think that would be useful here, because some operations do while others don't.

Reply:
We can possibly use an immutable `ReadonlyRequirementFulfillmentGraph` as the input graph. I think this can be tackled in a later PR, though.

Reply:
Added in #762